### PR TITLE
Add additional documentation for handling collections. Fix behavior of `waitForCollectionCallback`.

### PR DESCRIPTION
**Pull request by marcaaron**
### Details

Fixes issue found here: https://github.com/Expensify/App/pull/10051#issuecomment-1211421037

Semi-related to: https://github.com/Expensify/Expensify/issues/222064

### Automated Tests

I can add some but the ones that exist are passing.

### Linked PRs

TODO

### CODE REVIEW

File: README.md
Comment (luacmartins) [inline README.md:142]: NAB but should we use an existing key as an example? That would make it easier for a contributor to search for examples in the code base and see how collections are used.

```suggestion
Collections allow keys with similar value types to be subscribed together by subscribing to the collection key. To define one, it must be included in the `ONYXKEYS.COLLECTION` object and it must be suffixed with an underscore. Member keys should use a unique identifier or index after the collection key prefix (e.g. `report_42`).
```

File: README.md
Comment (dangrous) [inline README.md:186]: Where is `n` defined? I don't see it in the code above. Is it once per member key? Either way, might be good to clarify here. (Coming as someone with no context on this whatsoever.)

**Review by luacmartins — DISMISSED** [review]
LGTM left a small NAB. Thanks for fixing the bug and adding more docs on collections!

File: README.md
Comment (dangrous) [inline README.md:202]: Same thought as above - where is `n` coming from? If this is a silly comment, feel free to ignore me.

**Review by dangrous — COMMENTED** [review]
Knowing very little about the context of this yet, this seems well written and understandable. Added a couple of places that could use clarification.

As for the actual content of it, I'll defer to someone who knows what they're talking about.

File: README.md
Comment (marcaaron) [reply README.md:142]: Ah so, my intention here was to try to keep this generic so that people could imagine using it for whatever their use case and not get thrown off by "expensify" language. But I think the point gets across well enough here either way though so I'll update it.

File: README.md
Comment (marcaaron) [reply README.md:202]: Hey, if you didn't get it the first time then the docs could be better! So, not silly at all 😄

File: README.md
Comment (luacmartins) [reply README.md:142]: Makes sense! Thanks for the changes

File: README.md
Comment (marcaaron) [reply README.md]: Added an example to help clarify.

**Comment by dangrous** [comment]
Updates look great to me! Resolves my questions perfectly.

**Review by luacmartins — APPROVED** [review]

**Comment by tgolen** [comment]
I was sort of holding off on reviewing this since @marcaaron and I were still having an ongoing discussion about whether a parameter is better or a separate method is better. Also, we should consider changing this param/method to be the default behavior for collections.

@marcaaron Did you want to make some of those changes in this PR?

**Comment by marcaaron** [comment]
We can keep discussing and I can make them in a follow up (or just not at all). This PR largely just improves documentation and fixes a bug on `main` (and eventually `staging`).

File: README.md
Comment (tgolen) [inline README.md:160]: Can you please add a little more guidance here so people know which one to choose? I think adding information to say that when there are multiple keys, you should always use mergeCollect() because it optimizes all the UI updates. If it's a single key, it's fine to use either.

File: README.md
Comment (tgolen) [inline README.md:180]:
```suggestion
This will add a prop to the component called `allReports` which is an object of collection member key/values. Changes to the individual member keys will modify the entire object and new props will be passed with each individual key update. The prop doesn't update on the initial rendering of the component until the entire collection has been read out of Onyx.
```

File: README.md
Comment (tgolen) [reply README.md:160]: Ah, I see you added all that information below. Maybe just add something here that says "See below for guidance on best practices"

File: README.md
Comment (tgolen) [inline README.md:206]:
```suggestion
_.each(reports, report => Onyx.merge(`${ONYXKEYS.COLLECTION.REPORT}${report.reportID}`, report)); // -> A component using withOnyx() will have it's state updated with each iteration
```

File: README.md
Comment (tgolen) [inline README.md:211]:
```suggestion
Onyx.mergeCollection(ONYXKEYS.COLLECTION.REPORT, values); // -> A component using withOnyx() will only have it's state updated once
```

**Review by tgolen — CHANGES_REQUESTED** [review]

File: README.md
Comment (marcaaron) [reply README.md:180]: Nice!

**Comment by marcaaron** [comment]
Updated. Thanks for the notes @tgolen 🙇

**Review by tgolen — APPROVED** [review]

**Pull request merged**
